File: 209-minimum-size-subarray-sum/minimum-size-subarray-sum.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)
        prefix = [0] * (n + 1)
        for i in range(n):
            prefix[i + 1] = prefix[i] + nums[i]
        
        res = n + 1
        for i in range(n):
            L, R = i, n
            while L < R:
                m = (R + L) // 2
                curr_sum = prefix[m + 1] - prefix[i]
                if curr_sum >= target:
                    R = m
                else:
                    L = m + 1
            if L < n:
                res = min(res, L - i + 1)
        return res % (n + 1)

File: 209-minimum-size-subarray-sum/minimum-size-subarray-sum.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int i = 0, j = 0, curr = 0, res = INT_MAX;
        int n = nums.size();
        while (j < n) {
            curr += nums[j++];
            if (curr >= target) {
                while (curr >= target) {
                    curr -= nums[i];
                    ++i;
                }
                res = min(res, j - i + 1);
            }
        }
        return (res == INT_MAX) ? 0 : res;
    }
};